sdk/src/types/block/output/storage_score.rs

```diff
@@ -0,0 +1,182 @@
+        self.offset_staking_feature = offset;
+        self
+    }
+
+    /// Sets the storage score offset for delegation fields.
+    pub fn with_delegation_offset(mut self, offset: u64) -> Self {
+        self.offset_delegation = offset;
+        self
+    }
+
+    /// Returns the storage cost per unit of storage score.
+    pub fn storage_cost(&self) -> u64 {
+        self.storage_cost
+    }
+
+    /// Returns the storage score factor for data fields.
+    pub fn data_factor(&self) -> u8 {
+        self.factor_data
+    }
+
+    /// Returns the storage score offset overhead per output.
+    pub fn output_overhead_offset(&self) -> u64 {
+        self.offset_output_overhead
+    }
+
+    /// Returns the storage score offset for Ed25519 block issuer key fields.
+    pub fn ed25519_block_issuer_key_offset(&self) -> u64 {
+        self.offset_ed25519_block_issuer_key
+    }
+
+    /// Returns the storage score offset for staking fields.
+    pub fn staking_feature_offset(&self) -> u64 {
+        self.offset_staking_feature
+    }
+
+    /// Returns the storage score offset for delegation fields.
+    pub fn delegation_offset(&self) -> u64 {
+        self.offset_delegation
+    }
+
+    /// Returns the storage score offset per output.
+    pub fn output_offset(&self) -> u64 {
+        self.output_overhead_offset()
+            + (self.data_factor() as usize * (OutputId::LENGTH + BlockId::LENGTH + core::mem::size_of::<SlotIndex>()))
+                as u64
+    }
+
+    /// Returns the storage score offset for implicit account creation address fields.
+    pub fn implicit_account_creation_address_offset(&self) -> u64 {
+        let null_address = Ed25519Address::null();
+        let basic_output_score = BasicOutputBuilder::new_with_amount(0)
+            .add_unlock_condition(AddressUnlockCondition::new(null_address))
+            .storage_score(*self);
+        let ed25519_address_score = null_address.storage_score(*self);
+        let basic_score_without_address = basic_output_score - ed25519_address_score;
+        let account_output_score = AccountOutputBuilder::new_with_amount(0, AccountId::null())
+            .add_unlock_condition(GovernorAddressUnlockCondition::new(null_address))
+            .add_unlock_condition(StateControllerAddressUnlockCondition::new(null_address))
+            .add_feature(BlockIssuerFeature::new(0, [BlockIssuerKey::Ed25519(Ed25519BlockIssuerKey::null())]).unwrap())
+            .storage_score(*self);
+        account_output_score - basic_score_without_address
+    }
+}
+
+/// A trait to facilitate the computation of the byte cost of block outputs, which is central to dust protection.
+pub trait StorageScore {
+    fn storage_score(&self, _params: StorageScoreParameters) -> u64 {
+        0
+    }
+
+    /// Computes the storage cost of this instance given [`StorageScoreParameters`].
+    fn storage_cost(&self, params: StorageScoreParameters) -> u64 {
```

Comment:
I don't think we should call this `storage_cost` as the name is already reserved for one of the component of what this is actually returning. TIP calls this `Minimum Storage Deposit` so we should call it exactly the same.

Reply:
but that's not what it is...

Reply:
How so ?

Reply:
It's not the min storage deposit. it can be calculated for an individual unlock condition, for instance

Reply:
Ok so there is a single place where this is being used as non min_storage_deposit, in a test. I think I need to better understand what you are trying to achieve because it doesn't feel very correct to me. Storage score of a SDRUC is most of the time 0, only case would be if it contains an implicit account creation one but I'm not even sure this is possible, I need to ask iota-core.

Reply:
That being said, I don't see why it couldn't be called `Minimum Storage Deposit`, it's still really is a share of it, even if not an output.

Reply:
No, we use it to calculate the minimum required storage deposit, where we call it on a sdruc

Reply:
I think the term `Minimum Storage Deposit` is really only defined for the outputs (could be made clearer in the TIP tho), and rather not for its constituents, so I think we're good calling it however we like. However, in the protocol parameters a slightly more accurate term would have been: `StorageCostPerScore` to determine the individual "Storage Cost" of a constituent which is used to determine the "minimum storage deposit" of an output.

Reply:
My point is that currently `storage_cost` is used for only ONE thing that is not minimum storage deposit and I'm not convinced it's needed in the first place. But we can discuss that in the huddle.

Reply:
I leave it up to you guys to decide. My preference tho would be to just ask the TIP writers to rename `storage_cost` to `storage_cost_factor` on the `StorageScoreParameters` to remove the name collision and then leave it as it is. I kinda dislike renaming this to `minimum_storage_deposit` because 
1) if we ever wanted to use "storage cost" for something else we'ld have to rename this trait fn, while "storage_cost" is probably fairly future-proof.
2) it's uglier and who likes that? :trollface:

Reply:
1. I don't think we should ever account for future proofing names, this logic could apply to way more than this.
2. It's not uglier and everybody wants that, it's always been named like that, renaming it would create confusion.
3. I would do that